feat: Initial tool calling engine implementation

AtakanTekparmak opened the pull request:
- Added skeleton implementation for the `ToolCallingEngine` (everything except extracting the tool calls from LLM response is implemented, waiting for further discussion for that)
- Added necessary requirements
- Updated gitignore to include any and all `venv` folders
- Updated the relevant todo file ([llm_agents/todos/tool_calling_engine_todo.md](https://github.com/marketagents-ai/MarketAgents/compare/tool-calling-engine-1?expand=1#diff-5eecb998ee240b567c430fec249c7fbdcab83dd4d9c0e0b0a70a8272459445f6))

interstellarninja commented:
will merge once you have included the following:
- fork the tiny_func_engine to avoid dependency
- enable iterative function calling until `tool_calls` are found in the llm completion
- execute and pass results back to llm for summary for each `tool_call` as part of `tool` role
- return the final summary from `content` to pass back to caller agent

AtakanTekparmak commented:
> will merge once you have included the following:
> 
> * fork the tiny_func_engine to avoid dependency
> * enable iterative function calling until `tool_calls` are found in the llm completion
> * execute and pass results back to llm for summary for each `tool_call` as part of `tool` role
> * return the final summary from `content` to pass back to caller agent

Addressed remarks:
* fork the tiny_func_engine to avoid dependency
* execute and pass results back to llm for summary for each `tool_call` as part of `tool` role (the "tool" role part to be implemented on agent-side)

Remarks that need further clarification:
* enable iterative function calling until `tool_calls` are found in the llm completion: shouldn't this also be handled on agent-side (or possibly where LLM inference is made) because essentially it's a retry functionality?
* return the final summary from `content` to pass back to caller agent: So we want to return the formatted tool call results and unformatted LLM response (content) together back to the caller agent?

interstellarninja commented:
> Remarks that need further clarification:
> 
> * enable iterative function calling until `tool_calls` are found in the llm completion: shouldn't this also be handled on agent-side (or possibly where LLM inference is made) because essentially it's a retry functionality?

well when `tools` are present and we forward the messages to the tool calling engine and use `run_ai_tool_completion` function from `ai_utilities.py`
https://github.com/marketagents-ai/MarketAgents/blob/8825b011c98eeedd39a0b2ca87ed9ebdbaf298aa/llm_agents/aiutilities.py#L71

we do the following with the response:
- until tool calls are present we execute and pass `tool` results back to the same function recursively
- if the message doesn't contain `tool_calls` but just `content` we pass it back to agent as final response

> * return the final summary from `content` to pass back to caller agent: So we want to return the formatted tool call results and unformatted LLM response (content) together back to the caller agent?

yes we can pass all the message chain back to caller agent (for logging and memory) with final response which is `text` or `json output`
